src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertyAndValueDeserializer.cs

```diff
@@ -1093,116 +1093,135 @@ private ODataProperty ReadTopLevelPropertyImplementation(IEdmTypeReference expec
             object propertyValue = missingPropertyValue;
             var customInstanceAnnotations = new Collection<ODataInstanceAnnotation>();
 
-            string payloadTypeName = null;
-            if (this.ReadingComplexProperty(propertyAndAnnotationCollector, expectedPropertyTypeReference, out payloadTypeName))
+            // Check for the special top-level null marker 
+            if (this.IsTopLevelNullValue() && this.MessageReaderSettings.LibraryCompatibility == ODataLibraryCompatibility.Version6)

```

Comment:
>this.IsTopLevelNullValue() && this.MessageReaderSettings.LibraryCompatibility == ODataLibraryCompatibility.Version6 [](start = 16, length = 115)

What if this.IsTopLevelNullValue() && != ODataLibCompatibility6 here? Should we handle that case here, when we read the IsTopLevelNullValue?

Or, since the payload is unambiguous in what it's trying to convey, even if it's not 4.0 compatible, should we be more lax *on reading* and always support, rather than requiring the client to specify "I might be talking to a bonehead service that returned the OData 3.0 representation of a top-level null"?  In other words, are there any cases where we would tell the client that they shouldn't set LibraryCompatibility6 *on the reader*?